streaming/replication.c

```diff
@@ -0,0 +1,1933 @@
+static inline void replication_send_chart_definition_nolock(RRDSET *st);
+GAP* add_gap_data(GAPS *host_queue, GAP *gap);
+
+/********************************
+* Thread Initialization functions
+*********************************/ 
+static void replication_state_init(REPLICATION_STATE *state)
+{
+    debug(D_REPLICATION, "%s: Replication State Initialization", REPLICATION_MSG);
+    memset(state, 0, sizeof(*state));
+    state->buffer = cbuffer_new(1024, 1024*1024);
+    state->build = buffer_create(1);
+    state->socket = -1;
+    state->dim_past_data = (RRDDIM_PAST_DATA *)callocz(1, sizeof(RRDDIM_PAST_DATA));
+    state->dim_past_data->page = (void *)callocz(MEM_PAGE_BLOCK_SIZE, sizeof(char));
+    netdata_mutex_init(&state->mutex);
+}
+
+void replication_state_destroy(REPLICATION_STATE **state)
+{
+    REPLICATION_STATE *r = *state;
+    pthread_mutex_destroy(&r->mutex);
+    if(r->buffer)
+        cbuffer_free(r->buffer);
+    buffer_free(r->build);
+#ifdef ENABLE_HTTPS
+    if(r->ssl.conn){
+        SSL_free(r->ssl.conn);
+    }
+#endif
+    if(r->dim_past_data && r->dim_past_data->page)
+        freez(r->dim_past_data->page);
+    freez(r->dim_past_data);
+    freez(*state);
+    debug(D_REPLICATION, "%s: Replication state destroyed.", REPLICATION_MSG);
+}
+
+void replication_sender_init(RRDHOST *host){
+    if(!host || !host->replication){
+        error("%s: Host or host's replication state is not initialized! - Replication Tx thread Initialization failed!", REPLICATION_MSG);
+        return;
+    }
+    host->replication->tx_replication = (REPLICATION_STATE *)callocz(1, sizeof(REPLICATION_STATE));
+    replication_state_init(host->replication->tx_replication);
+    host->replication->tx_replication->host = host;
+    host->replication->tx_replication->enabled = default_rrdpush_sender_replication_enabled;
+#ifdef ENABLE_HTTPS
+    host->replication->tx_replication->ssl.conn = NULL;
+    host->replication->tx_replication->ssl.flags = NETDATA_SSL_START;
+#endif
+    info("%s: Initialize Replication Tx thread state at host creation %s .", REPLICATION_MSG, host->hostname);
+}
+
+static unsigned int replication_rd_config(RRDHOST *host, struct config *stream_config, char *key)
+{
+    REPLICATION_STATE *rep_state = host->replication->rx_replication;
+    debug(D_REPLICATION, "%s: Reading configuration for Replication Rx thread for host %s ", REPLICATION_MSG, host->hostname);
+    
+    //TODO: Add any missing configuration.
+    rep_state->timeout = (int)appconfig_get_number(stream_config, CONFIG_SECTION_STREAM, "timeout seconds", 60);
+    rep_state->default_port = (int)appconfig_get_number(stream_config, CONFIG_SECTION_STREAM, "default port", 19999);    
+
+    unsigned int rrdpush_receiver_replication_enable = default_rrdpush_receiver_replication_enabled;
+    rrdpush_receiver_replication_enable = appconfig_get_boolean(stream_config, key, "enable replication", rrdpush_receiver_replication_enable);
```

Comment:
Any particular reason why a user would want to disable replication? I'm thinking that in the grand scale of things, it should be there helping streaming all the time.. ?

Reply:
My guess: we need this until this feature becomes stable, so a user has an option to disable it if there are any problems.

Reply:
@MrZammler I agree with the big picture you suggest. @ilyam8 made a hit, let's keep the enable configuration until this feature is robust enough to work assistive with streaming.